### DISCUSSION TITLE
Undo/Redo Support

### DISCUSSION THREAD

**Discussion by rickshane** [guoyingtao/Mantis#377] [Q&A] [answered]
What would be the easiest way to handle support for undo/redo?

I tried saving the Mantis.Config object from an existing CropViewController object and then passing it in to Mantis.setupCropViewController() on an existing CropViewController object and also to a new version in the Mantis.cropViewController init call. But this does not update the UI for the previous state. Is there a single function that will rebuild the UI based on the current state in the config object?

Thanks,

Rick

**Comment by guoyingtao** [comment]
Currently Mantis does not save view status into the config object, so it doesn't support undo/redo as you mentioned.

**Comment by rickshane** [comment]
I have undo/redo support working for all transformations except 90 degree rotations:

I took the logic from CropView.processPresetTransformation() for the .presetInfo case, and created a function for the CropViewProtocol called applyTransform() which contains this logic. 

Then I take a snapshot of the Transformation object before and after a transformation by calling cropView.makeTransformation(), and send both snapshots back to my delegate. I can then undo/redo the UI state by calling back in to a new cropViewController delegate which will call the applyTransform() function. 

It can successfully undo/redo the UI state for all transformations except a 90 degree rotation. Both undo and redo of a 90 degree rotation transform fail. 

It works for:

1. Aspect Ratio Change
2. Degree Rotation
3. Flip Horizontal
4. Flip Vertical
5. Crop by changing crop box
6. Crop by panning
7. Crop by zooming

func applyTransform(byTransformInfo transformation: Transformation, isUpdateRotationControlView: Bool = true) {
        
        viewModel.horizontallyFlip = transformation.horizontallyFlipped
        viewModel.verticallyFlip = transformation.verticallyFlipped
        
        if transformation.horizontallyFlipped {
            flipOddTimes.toggle()
        }
        
        if transformation.verticallyFlipped {
            flipOddTimes.toggle()
        }
        
        var newTransform = getTransformInfo(byTransformInfo: transformation)
        
        // The first transform is just for retrieving the final cropBoxFrame
        transform(byTransformInfo: newTransform, isUpdateRotationControlView: isUpdateRotationControlView)
        
        // The second transform is for adjusting the scale of transformInfo
        let adjustScale = (viewModel.cropBoxFrame.width / viewModel.cropBoxOriginFrame.width)
        / (transformation.maskFrame.width / transformation.initialMaskFrame.width)
        
        newTransform.scale *= adjustScale
        
        transform(byTransformInfo: newTransform)
    }

Any ideas what is missing? Is it the rotation type or the rotations in the viewModel?

**Comment by rickshane** [comment]
I just added the ImageRotationType state to the Transformation object. If I then set the rotationType in the viewModel from the new applyTransform() function, I can successfully undo a 90 degree rotation. 

Redo of a 90 degree rotation still fails.

**Comment by rickshane** [comment]
func applyTransform(byTransformInfo transformation: Transformation, isUpdateRotationControlView: Bool = true) {
        
        viewModel.rotationType = transformation.rotationType
        
        viewModel.horizontallyFlip = transformation.horizontallyFlipped
        viewModel.verticallyFlip = transformation.verticallyFlipped
        
        if transformation.horizontallyFlipped {
            flipOddTimes.toggle()
        }
        
        if transformation.verticallyFlipped {
            flipOddTimes.toggle()
        }
        
        var newTransform = getTransformInfo(byTransformInfo: transformation)
        
        // The first transform is just for retrieving the final cropBoxFrame
        transform(byTransformInfo: newTransform, isUpdateRotationControlView: isUpdateRotationControlView)
        
        // The second transform is for adjusting the scale of transformInfo
        let adjustScale = (viewModel.cropBoxFrame.width / viewModel.cropBoxOriginFrame.width)
        / (transformation.maskFrame.width / transformation.initialMaskFrame.width)
        
        newTransform.scale *= adjustScale
        
        transform(byTransformInfo: newTransform)
    }

**Comment by guoyingtao** [comment]
@rickshane 
Would you mind posting your project with undo/redo features here? Maybe a zip file? I can do some tests with that to see why redo a 90 degree fails.

**Comment by rickshane** [comment]
[Mantis_Undo_Redo.zip](https://github.com/guoyingtao/Mantis/files/14625780/Mantis_Undo_Redo.zip)


Here is a zip file of the project. 

Run the xcode project from the /Examples folder and select the "Embedded" version of the cropViewController. 

You should see yellow "Undo" and "Redo" buttons on the left side of the toolbar.

Thanks,

Rick

**Comment by rickshane** [comment]
It also looks like undoing an Aspect Ratio change back to "Freeform" (unlocked) will not remember the aspect ratio that was set prior to changing to the "Freeform" (unlocked) . Some additional state for that is needed somewhere to match the crop rect frame.

**Reply by guoyingtao** [reply]
Restoring image crop from Transformation is always considered `Freeform` currently. We may need to save more state information into Transformation for undo/redo

**Comment by guoyingtao** [comment]
@rickshane 
Please try `viewModel.rotationType = .none` instead of `viewModel.rotationType = transformation.rotationType` in applyTransform
The rotation value in Transformation already includes the effect of rotationType

**Comment by rickshane** [comment]
That doesn't work for redo (I get same results as before).

I also tried setting "viewModel.rotationType = .none" at beginning of applyTransform() and then "viewModel.rotationType = transformation.rotationType" as the last line of applyTransform() after the two calls to "transform()".

The rotationType state will still need to be restored, becase subsequent transforms sometimes look at rotationType.isRotatedByMultiple180

**Reply by guoyingtao** [reply]
Below is the code I am using. It works OK after changing `viewModel.rotationType = transformation.rotationType` to `viewModel.rotationType = .none`
<img width="859" alt="image" src="https://github.com/guoyingtao/Mantis/assets/26723384/4350aaec-b5a6-4884-bac4-97f0b0669329">

https://github.com/guoyingtao/Mantis/assets/26723384/dccc9001-05fc-42d1-b9a0-3a5cb69e6e43

**Reply by rickshane** [reply]
Does it work in Catalyst for you?

**Reply by rickshane** [reply]
It works on iPhone Portrait mode for me.

It fails on iPad Landscape and Catalyst.

**Reply by rickshane** [reply]
Landscape and Portrait mode on iPad fails
Landscape mode on iPhone 12 Pro Max fails

**Reply by guoyingtao** [reply]
I noticed that too. I haven't figured out the reason yet.

**Comment by guoyingtao** [comment]
@rickshane 
Can you comment out the code below to see if it works. It is for applying transform data to a different device, redo/undo happens on the same device, looks like we don't need them. 
I don't know why adding them causes the issue, but looks like it works OK without this **adjustScale**.


```Swift
        // The second transform is for adjusting the scale of transformInfo
        let adjustScale = (viewModel.cropBoxFrame.width / viewModel.cropBoxOriginFrame.width)
        / (transformation.maskFrame.width / transformation.initialMaskFrame.width)
        
        newTransform.scale *= adjustScale
```

**Reply by rickshane** [reply]
That's much better. I still add "viewModel.rotationType = transformation.rotationType" to restore the rotationType state so I can apply other transforms after redoing a 90 degree rotation.

The only outstanding issue seems to be undoing an aspect ratio change. I think this is because it starts in (free) unlocked without a ratio and then if you apply a new ratio and then undo back to the unlocked (free) version, the ratio does not get reset to what the ratio was when it was initially unlocked.

Thanks for spending time on this. I believe it is a very important feature.

**Reply by guoyingtao** [reply]
I agree the feature looks very useful.
For undoing an aspect ratio change, we may need save aspect operations along with transformation data. Maybe we need a operation status like
```Swift
CropOperationStatus {
   var transformation
   var aspectType
}
```
Then apply aspectType when undo/redo

I need another time to think about it but you are welcome to modify the code in your way to see if it works

**Reply by rickshane** [reply]
My Transformation type now includes aspect ratio and lock status, which seems to fix the problem with aspect ratio undo/redo. I also changed "rotations" to "radians" and don't include rotationType radians in that number. I then make the necessary changes in applyTransform() and transform() to account for the two types of radian changes.

public typealias Transformation = (
    offset: CGPoint,
    radians: CGFloat,
    scale: CGFloat,
    isManuallyZoomed: Bool,
    initialMaskFrame: CGRect,
    maskFrame: CGRect,
    cropWorkbenchViewBounds: CGRect,
    horizontallyFlipped: Bool,
    verticallyFlipped: Bool,
    rotationType: ImageRotationType,
    aspectRatioLockEnabled: Bool,
    aspectRato: CGFloat
)

**Comment by rickshane** [comment]
I will also need to store flipOddTimes to correctly undo a flip that occurred after a rotation. I will then work on refactoring into new CropState object that stores Transformation plus all the other state needed for undo/redo. 

But should rotationType stay in the Transformation object along with radians?

**Reply by guoyingtao** [reply]
I think `rotationType `, `aspectRatioLockEnabled `, and `aspectRatio` can all be moved out of Transformation. For me they are part of cropping status but not part of Transformation.

**Reply by rickshane** [reply]
Here is an updated version of the project with new CropState data structure
[Mantis_Undo_Redo.zip](https://github.com/guoyingtao/Mantis/files/14646714/Mantis_Undo_Redo.zip)

**Reply by guoyingtao** [reply]
Looks nice!

**Comment by rickshane** [comment]
I notice there is a big bug in the control that is affecting Undo/Redo logic. If I touch down just outside the crop rectangle, it causes the grid to move, even though I have not moved my finger. At a high level, this always prevents pulling the crop grid to the edge of the image. This is the biggest bug I have noticed since using this control. 

This affects undo/redo because if I click just outside the lower right anchor point, it causes handle to move and inside the function "imageStatusChanged" there is a check on the point, "getImageRightBottomAnchorPoint", and since this moved, this function returns true which causes a new transform record to be pushed on to the undo stack.

**Reply by guoyingtao** [reply]
I think we can check if the previous state equals the current one like below
```Swift
        guard let previous = previousCropState else { return }
        let userGenerated = isCropStateUserGenerated
        currentCropState = cropView.makeCropState()
        
        if previous != currentCropState {
            delegate?.cropViewControllerDidTransformImage(previous: previous, current: currentCropState, userGenerated: userGenerated)
        }
```

For convenience, we also need to change CropState and Transformation to struct and make them Equatable
```Swift
public struct CropState: Equatable {
    var rotationType: ImageRotationType
    var degrees: CGFloat
    var aspectRatioLockEnabled: Bool
    var aspectRato: CGFloat
    var flipOddTimes: Bool
    var transformation: Transformation
    
    public init(rotationType: ImageRotationType, degrees: CGFloat, aspectRatioLockEnabled: Bool, aspectRato: CGFloat, flipOddTimes: Bool, transformation: Transformation) {
        self.rotationType = rotationType
        self.degrees = degrees
        self.aspectRatioLockEnabled = aspectRatioLockEnabled
        self.aspectRato = aspectRato
        self.flipOddTimes = flipOddTimes
        self.transformation = transformation
    }
    
    public static func ==(lhs: CropState, rhs: CropState) -> Bool {
        return lhs.rotationType == rhs.rotationType
        && lhs.degrees == rhs.degrees
        && lhs.aspectRatioLockEnabled == rhs.aspectRatioLockEnabled
        && lhs.aspectRato == rhs.aspectRato
        && lhs.flipOddTimes == rhs.flipOddTimes
        && lhs.transformation == rhs.transformation
    }
}

public struct Transformation: Equatable {
    var offset: CGPoint
    var rotation: CGFloat
    var scale: CGFloat
    var isManuallyZoomed: Bool
    var initialMaskFrame: CGRect
    var maskFrame: CGRect
    var cropWorkbenchViewBounds: CGRect
    var horizontallyFlipped: Bool
    var verticallyFlipped: Bool
    
    public init(offset: CGPoint, rotation: CGFloat, scale: CGFloat, isManuallyZoomed: Bool, initialMaskFrame: CGRect, maskFrame: CGRect, cropWorkbenchViewBounds: CGRect, horizontallyFlipped: Bool, verticallyFlipped: Bool) {
        self.offset = offset
        self.rotation = rotation
        self.scale = scale
        self.isManuallyZoomed = isManuallyZoomed
        self.initialMaskFrame = initialMaskFrame
        self.maskFrame = maskFrame
        self.cropWorkbenchViewBounds = cropWorkbenchViewBounds
        self.horizontallyFlipped = horizontallyFlipped
        self.verticallyFlipped = verticallyFlipped
    }
    
    public static func ==(lhs: Transformation, rhs: Transformation) -> Bool {
        return lhs.offset == rhs.offset
        && lhs.rotation == rhs.rotation
        && lhs.scale == rhs.scale
        && lhs.isManuallyZoomed == rhs.isManuallyZoomed
        && lhs.initialMaskFrame == rhs.initialMaskFrame
        && lhs.maskFrame == rhs.maskFrame
        && lhs.cropWorkbenchViewBounds == rhs.cropWorkbenchViewBounds
        && lhs.horizontallyFlipped == rhs.horizontallyFlipped
        && lhs.verticallyFlipped == rhs.verticallyFlipped
    }
}
```

**Reply by rickshane** [reply]
The comparison test still fails because the maskFrame value has changed, when clicking just outside the frame which causes it to shrink from the right.

**Reply by guoyingtao** [reply]
If only clicking just outside the frame, it works for me. (I only tested on simulators)
Would it be possible to upload a video to help me understand the issue better? Thank you!

**Reply by rickshane** [reply]
I am testing in Catalyst. I just verified the bug does not occur on an iOS device. Only in Catalyst. 

Please test in Catalyst.

**Reply by guoyingtao** [reply]
I just tested in Catalyst on my laptop and it worked too. My Mac system is Sonoma 14.3.1
What is your Mac system version?

**Reply by rickshane** [reply]
MacBook Air M2
Sonoma 14.1.2

**Reply by rickshane** [reply]
https://github.com/guoyingtao/Mantis/assets/5288729/1b1352ff-05f7-48b7-bcd3-4c10f52e8582

**Reply by rickshane** [reply]
And here is another video which shows that if you click outside the grid and then resize, you can never pull the grid to the edge of the image (this is the high level bug I mentioned that I think is a much bigger issue, although it is the same bug.)

https://github.com/guoyingtao/Mantis/assets/5288729/9945b761-1f3b-4ec7-af5b-0847ceb155ba

**Reply by guoyingtao** [reply]
Below is the video of my test in Catalyst
https://github.com/guoyingtao/Mantis/assets/26723384/c95ea8a4-c3be-4bc2-963d-94c887109bfd

**Reply by guoyingtao** [reply]
OK. I saw the issue. 
I will go back to check it when I get more time.

**Reply by guoyingtao** [reply]
@rickshane 
I figured it out.
For Mac Catalyst, even you just click a spot without moving, it still trigged `touchesMoved` which is different on iPhone/iPad
You can change ` updateCropBoxFrame(withTouchPoint: touchPoint)` of `touchesMoved` in `CropView+Touches` to the code below, it should solve the issue

```Swift
if touchPoint != viewModel.panOriginPoint {
    updateCropBoxFrame(withTouchPoint: touchPoint)
}
```

**Reply by rickshane** [reply]
This is better. There is still a Catalyst touch bug in the Rotation Dial that affects undo/redo results. When you move the dial to the limit in Catalyst, if you keep going, it inverts to the negative value. 

Here is a video:


https://github.com/guoyingtao/Mantis/assets/5288729/01c969ff-68fb-4319-86b5-87c9620b1985

**Reply by guoyingtao** [reply]
@rickshane 
Can you add this code into `private func updateRotation(bySteppingAngle:)` in RotationDial after `let radians = steppingAngle.radians` to see if it solved issue?

```Swift
        guard radians != 0 else {
            return false
        }
```

**Reply by rickshane** [reply]
That seems to fix it. Thanks! 

I am still seeing the high level bug that prevents pulling grid to edge of image in catalyst. Here is an updated video:


https://github.com/guoyingtao/Mantis/assets/5288729/f38bf08c-5b0b-4173-b54b-055d2b4badeb

**Reply by guoyingtao** [reply]
@rickshane 
I haven't found a good solution to fix this issue. (If you do it slower, it seems fine)

**Comment by rickshane** [comment]
[Mantis_Undo_Redo.zip](https://github.com/guoyingtao/Mantis/files/14657510/Mantis_Undo_Redo.zip)

Here is the new version of the project with the Equatable struct versions of CropState and Transformation.

**Reply by guoyingtao** [reply]
Looks good. If it works most of times, maybe we can consider to integrate this feature to Mantis.

**Comment by rickshane** [comment]
[Mantis_Undo_Reset.zip](https://github.com/guoyingtao/Mantis/files/14666079/Mantis_Undo_Reset.zip)


This latest version has support for undo/redo of the "Reset" button. There are now two different types of Transform records pushed on the stack - "Transform" and "ResetTransforms". It also disables/enables the undo/redo/reset buttons based on the state of the UndoManager and Transform Stack.

I have also hooked up the "Undo" + "Redo" + "Revert to Original" menu commands in Catalyst using the AppDelegate.buildMenu function.

**Reply by guoyingtao** [reply]
Great!
BTW, when doing `reset`, I think we need to clear undo/redo stack.

**Reply by rickshane** [reply]
Only clear undo/redo stack when enter or exit crop mode. This is why there are two types of records, "Transform" and "ResetTransforms". You need to be able to undo/redo a "Reset", therefore that record needs to be present on the stack.

**Comment by rickshane** [comment]
This version has the fix for Catalyst touchesMoved bug. I have added clockwiseRotate, and both flip buttons to tool bar for testing undo/redo.

[Mantis_Catalyst_Touch_Fix.zip](https://github.com/guoyingtao/Mantis/files/14673083/Mantis_Catalyst_Touch_Fix.zip)

**Comment by rickshane** [comment]
I just sent a pull request for this feature.

**Comment by rickshane** [comment, marked as answer]
I will close this as "Answered".
